### PR TITLE
[CWS][SEC-106] Add ability to identify interpreter of a script

### PR DESCRIPTION
**Pull request by modernplumbing**
### What does this PR do?
Adds ability to identify the interpreter of a script inside a script via the shebang. This is currently limited to one layer of nested script. Example rule would be `exec.interpreter.file.name == ~"python*"`

### Motivation

Currently, when a script like the below is run, we can identify that it was a bash script, but cannot identify that a python script was also run. This PR adds the ability to identify that Python was run. 
```
`#!/bin/bash

echo "Executing echo inside a bash script"

cat << EOF > pyscript.py
#!/usr/bin/python3

print('Executing print inside a python script')

EOF

echo "Back to bash"

chmod 755 pyscript.py
./pyscript.py`
```

### Additional Notes

The current implementation does not expose the interpreter of any script that's nested more than 1 layer. For example, in the following script, `perl` cannot be identified at the rule level yet.

```
echo "Executing echo inside a bash script"

cat << '__HERE__' > hello.pl
#!/usr/bin/perl

my $foo = "Hello from Perl";
print "$foo\n";

__HERE__

chmod 755 hello.pl

cat << EOF > pyscript.py
#!/usr/bin/python3

import subprocess

print('Executing print inside a python script')

subprocess.run(["perl", "./hello.pl"])

EOF

echo "Back to bash"

chmod 755 pyscript.py
./pyscript.py
```

### Possible Drawbacks / Trade-offs

Because I'm only sending the inode, mount id, and path id to userspace, the documented fields like 

```
| `exec.interpreter.file.rights` | int | Mode/rights of the file | Chmod mode constants |
| `exec.interpreter.file.uid` | int | UID of the file's owner |  |
| `exec.interpreter.file.user` | string | User of the file's owner |  |
```
are actually not available. This can be confusing for the user, so i'm considering sending the whole file information, even if it adds to the size of the data.

### Describe how to test/QA your changes

See `TestProcessIdentifyInterpreter` for examples.

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Comment by sgnn7** [comment]
@modernplumbing I know it's a draft PR but do you mind updating the title of this PR to indicate what it's trying to do? Since it's a live PR, everyone watching this repo is being notified of its pushes which come in titled with the current PR title.

**Comment by modernplumbing** [comment]
> @modernplumbing I know it's a draft PR but do you mind updating the title of this PR to indicate what it's trying to do? Since it's a live PR, everyone watching this repo is being notified of its pushes which come in titled with the current PR title.

Sorry about that!

File: pkg/security/ebpf/c/exec.h [hunk 1]
```diff
@@ -597,6 +666,38 @@ int kprobe_security_bprm_check(struct pt_regs *ctx) {
...
 }
 
+SEC("kprobe/setup_new_exec")
+int kprobe_setup_new_exec(struct pt_regs *ctx) {
+    struct syscall_cache_t *syscall = peek_syscall(EVENT_EXEC);
+    if (!syscall) {
+        return 0;
+    }
+
+    struct linux_binprm *bprm = get_linux_binprm_obj();
```
Comment (lebauce) [inline pkg/security/ebpf/c/exec.h:676]: Why fetching it from a map and override it with an argument ?

Comment (modernplumbing) [reply pkg/security/ebpf/c/exec.h:684]: ah, this is cruft I haven't deleted yet. thanks for catching it!

**Review by brycekahle — APPROVED** [review]
👍 for `tasks/system_probe.py` change

**Review by alai97 — APPROVED** [review]
Looks good for docs! There are some merge conflicts that need to be resolved.

File: pkg/security/tests/module_tester.go [hunk 2]
```diff
@@ -316,6 +316,21 @@ func which(tb testing.TB, name string) string {
 	return executable
 }
 
+//nolint:deadcode,unused
+// whichNotFatal returns an error instead of fatal
+func whichNotFatal(tb testing.TB, name string) (string, error) {
```
Comment (lebauce) [inline pkg/security/tests/module_tester.go:321]: What about making the `which` function call the `whichNotFatal` function and make the t.Fatalf ?

File: pkg/security/probe/probe.go [hunk 3]
```diff
@@ -1492,6 +1492,11 @@ func NewProbe(config *config.Config, statsdClient statsd.ClientInterface) (*Prob
 			Name:  "syscall_monitor_event_period",
 			Value: uint64(p.config.ActivityDumpSyscallMonitorPeriod.Nanoseconds()),
 		},
+		manager.ConstantEditor{
+			Name: "setup_new_exec_is_last",
+			// TODO!!!
```
[the remark this answers is not in the document]
Comment (modernplumbing) [reply pkg/security/probe/probe.go:1497]: addressed!

File: pkg/security/probe/process_resolver.go [hunk 4]
```diff
@@ -421,6 +421,34 @@ func (p *ProcessResolver) enrichEventFromProc(entry *model.ProcessCacheEntry, pr
...
+	//print "Hello from Perl\n";
+	//
+	//EOF
+	if valueCount := len(entry.ArgsEntry.Values); valueCount > 1 {
+		firstArg := entry.ArgsEntry.Values[0]
+		lastArg := entry.ArgsEntry.Values[valueCount-1]
+		// Example result: comm value: pyscript.py | args: [/usr/bin/python3 ./pyscript.py]
+		if path.Base(lastArg) == entry.Comm && path.IsAbs(firstArg) {
+			entry.LinuxBinprm.FileEvent = entry.FileEvent
+			seclog.Debugf("Likely interpreter event detected in process cache. Pathname: %s", entry.LinuxBinprm.FileEvent.PathnameStr)
```
[the remark this answers is not in the document]
Comment (modernplumbing) [reply pkg/security/probe/process_resolver.go:448]: removed!

File: pkg/security/tests/module_tester.go (hunk 2 above)
Comment (modernplumbing) [reply pkg/security/tests/module_tester.go]: good catch!

File: releasenotes/notes/runtime-security-identify-interpreter-6f534194ba19d4f0.yaml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+# Each section from every release note are combined when theª
```
Comment (lebauce) [inline releasenotes/notes/runtime-security-identify-interpreter-6f534194ba19d4f0.yaml:1]: nit: remove comments

**Review by lebauce — APPROVED** [review]

**Review by paulcacheux — APPROVED** [review]

**Pull request merged**